### PR TITLE
fix(readBody): handle top-level arrays in url-encoded data

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

resolves https://github.com/unjs/h3/issues/275

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Previously a valid url-encoded string with multiple entries for the same value was not properly parsed. This creates arrays in such cases; it still doesn't support advanced nested arrays like `qs` but handles this specification as expected.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/utils/body.ts [hunk 1]
```diff
@@ -64,7 +64,16 @@ export async function readBody<T=any> (event: H3Event): Promise<T> {
   const body = await readRawBody(event) as string;
 
   if (event.node.req.headers["content-type"] === "application/x-www-form-urlencoded") {
-    const parsedForm = Object.fromEntries(new URLSearchParams(body));
+    const form = new URLSearchParams(body);
+    const parsedForm: Record<string, any> = {};
```
[the remark this answers is not in the document]
Comment (danielroe) [reply src/utils/body.ts:68]: Out of curiosity, why?

File: src/utils/body.ts [hunk 2]
```diff
@@ -64,7 +64,16 @@ export async function readBody<T=any> (event: H3Event): Promise<T> {
...
+    for (const [key, value] of form.entries()) {
+      if (key in parsedForm && !Array.isArray(parsedForm[key])) {
+        parsedForm[key] = [parsedForm[key]];
+        parsedForm[key].push(value);
```
Comment (pi0) [inline src/utils/body.ts:72]: I think we need to handle when key is there and value is already an array (to test, adding another prop with same name)

File: src/utils/body.ts [hunk 3]
```diff
@@ -64,7 +64,16 @@ export async function readBody<T=any> (event: H3Event): Promise<T> {
...
+      } else {
+        parsedForm[key] = value;
```
Comment (pi0) [inline src/utils/body.ts:74]: Style: This behavior can be short circuit before array check.

File: src/utils/body.ts (hunk 2 above)
Comment (danielroe) [reply src/utils/body.ts:72]: I don't think we support arrays at all, at the moment. So I think this is safe.

Comment (danielroe) [reply src/utils/body.ts:72]:
```js
[...new URLSearchParams('a[]=j').entries()]
// [ "a[]", "j" ]
[...new URLSearchParams('a=[j]').entries()] 
// // [ "a", "[j]" ]
```

File: src/utils/body.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pi0) [reply src/utils/body.ts:68]: Possible prototype pollution. This makes constructor null (we might have missed some other places)

File: src/utils/body.ts (hunk 3 above)
Comment (danielroe) [reply src/utils/body.ts:74]: if you wish we can use `continue` but I'm not sure it's clearer

Comment (pi0) [reply src/utils/body.ts:76]: indeed without seems better

**Review by pi0 — APPROVED** [review]

**Comment by alphaelf** [comment]
on [this line](https://github.com/unjs/h3/blob/03243bab1c5db92ec615fff073101114e104103c/src/utils/body.ts#L98) and [this line](https://github.com/unjs/h3/blob/03243bab1c5db92ec615fff073101114e104103c/src/utils/body.ts#L100) can we use `destr` to store typed value instedof string value?, for now the `undefined, true, false, null, ...` values will returned as string

**Pull request merged**